### DISCUSSION TITLE
How does madsim override libc functions?

### DISCUSSION THREAD

**Discussion by xxchan** [madsim-rs/madsim#159] [Q&A]
I only noticed code like this. How does it work? Is there any magic? :eyes:

https://github.com/madsim-rs/madsim/blob/0d6a86ae3153d2c576a40bfb8e657ded92fda385/madsim/src/sim/time/system_time.rs#L3-L6

**Comment by xxchan** [comment]
Oh! Is it sth like "strong symbol" vs "weak symbol"?

**Reply by wangrunji0408** [reply]
Right! The strong symbol we defined would override the weak symbol defined in libc.

**Comment by xxchan** [comment]
~~And does this mean that adding madsim to the dependency will immediately override it? (Does this introduce fixed overhead? Or is it only replaced inside madsim crate?) 🤔~~ (Oh, I think it will replace when 1. adding madsim as dependency 2. program is compiled with `RUSTFLAGS="--cfg madsim"`)

Another question is what will happen if I have the same override in my program? 😄 

(I might need to relearn how linking works...)

**Reply by wangrunji0408** [reply]
> Another question is what will happen if I have the same override in my program?

If there are multiple strong symbols, the linker will raise an error. 😄
